src/program/latex/diagram/wire.rs

```diff
@@ -0,0 +1,198 @@
+        let mut dagger_count = 0;
+        for modifier in gate.modifiers {
+            match modifier {
+                // return error for unsupported modifier FORKED
+                GateModifier::Forked => {
+                    return Err(LatexGenError::UnsupportedModifierForked);
+                }
+                GateModifier::Dagger => {
+                    dagger_count += 1;
+                }
+                GateModifier::Controlled => {
+                    ctrl_count += 1;
+                }
+            }
+        }
+
+        // return the QuantikzGate form of Gate
+        Ok(QuantikzGate {
+            name: canonical_gate,
+            dagger_count,
+            ctrl_count,
+        })
+    }
+}
+
+/// The type of cell that can be stored in a column on the wire.
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
+pub(crate) enum QuantikzCellType {
+    #[default]
+    Empty,
+    Gate(QuantikzGate),
+    Ctrl(i64),
+}
+
+/// A single column on the wire containing cells and associated parameters.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct QuantikzColumn {
+    /// a column on the wire containing renderable items
+    pub(crate) cell: QuantikzCellType,
+    /// the Parameter on the wire at some column
+    pub(crate) parameter: Parameter,
+}
+
+impl QuantikzColumn {
+    /// Retrieves a gate's parameters from Expression and matches them with its
+    /// symbolic definition which is then stored into wire at the specific
+    /// column.
+    ///
+    /// # Arguments
+    /// `expression` - expression from Program to get name of Parameter
+    /// `texify` - is texify_numerical_constants setting on?
+    pub(crate) fn set_param(&mut self, expression: &Expression, texify: bool) {
+        // get the name of the supported expression
+        let text = match expression {
+            Expression::Address(mr) => mr.name.to_string(),
+            Expression::Number(c) => c.re.to_string(),
+            expression => expression.to_string(),
+        };
+
+        // if texify_numerical_constants
+        let param = if texify {
+            // set the texified symbol
+            Parameter::Symbol(Symbol::from_str(&text).unwrap_or(Symbol::Text(text)))
+        } else {
+            // set the symbol as text
+            Parameter::Symbol(Symbol::Text(text))
+        };
+
+        self.parameter = param;
+    }
+}
+
+/// A Wire represents a single qubit. This is a row vector, or [1 x n] matrix,
+/// where n, is the total number of Quil instructions (or columns). Each column
+/// on the wire maps to some item that can be serialized into LaTeX using the
+/// ``Quantikz`` RenderCommands. A wire is part of the Circuit which is an [m x
+/// n] matrix where m, is the total number of wires, or length, of the Circuit.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct Wire {
+    /// the columns on the wire that can be serialized into LaTeX
+    pub(crate) columns: Vec<QuantikzColumn>,
+}
```

Comment:
I feel like there is a conflict with the naming here. The name `Wire` (and docstring) implies it should represent a single row/qubit but it actually contains a vector of columns.

I think "column" is the right abstraction here, that is, each instruction in the program corresponds to a single column in the diagram. The docstring hints at this kind of structure, but it's not modeled by the data structures being used.

Sometimes it's helpful to start with the data structures, then build an implementation on top of that. What if we started with these:

```rust
struct Diagram {
    settings: RenderSettings,
    wires: HashMap<usize, usize>, // qubit index -> column index
    columns: Vec<DiagramColumn>
}

struct DiagramColumn {
    cells: Vec<QuantikzCell>
}

struct QuantikzCellType {
    Gate(QuantikzGate),
    Control(i64),
    Target,
    Empty,
}

struct QuantikGate {
    name: String,
    parameters: Vec<Parameter>,
    dagger_count: usize,
}
```

At a glance, all of these names feel a bit more clear to me. We have a diagram, which is composed of columns, and each column is composed of individual cells with a quantikz command. We also store the mapping of the qubit indices to their corresponding column index.

An API build on top of this might look something like:

```rust
impl Diagram {
    fn new(qubits: Vec<int>, settings: RenderSettings) -> Self {
         // Initialize a new `Diagram` with the given qubits
         // If settings.impute_missing qubits, then pad the gaps.
    }

    fn add_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
        // Initialize a column full of empty cells to be filled in.
        let mut column: Vec<QuantikzColumn> = vec![QuantikzCellType::Empty; len(self.wires)] 
        
        // This is oversimplified for the example. There will be a little
        // bit of pre-processing necessary to get your "canonical gate",
        // and it probably makes sense to go in reverse order so you can start
        // with the target qubit.
        for each qubit in gate.qubits.map(|q| q.into_fixed())? {
            column_index = self.wires.get(qubit)
            // set column[column_index] to the correct QuantikzCellType
        }

        self.columns.push(column)
    }
}
```

Does this make sense as an outline? Is it clear how `to_latex` would interact with `Diagram`?